### PR TITLE
Progress output see #98

### PR DESCRIPTION
**Pull request by micheleorselli**
The PR is still a WIP (fixes https://github.com/phparkitect/arkitect/issues/98) , though enough to get feedback, comments below
There is a major change about the runner, switching from a callback based approach to a plain old iterative way
The code still needs to be cleaned up

### CODE REVIEW

File: src/Analyzer/FileVisitor.php [hunk 1]
```diff
@@ -45,12 +47,24 @@ public function onClassAnalyzed(callable $callable): void
...
+        $this->classDescriptions = [];
+    }
+
     public function leaveNode(Node $node): void
     {
         if ($node instanceof Node\Stmt\Class_) {
             $classDescription = $this->classDescriptionBuilder->get();
 
-            \call_user_func($this->callable, $classDescription);
+            $this->classDescriptions[] = $classDescription;
```
Comment (micheleorselli) [inline src/Analyzer/FileVisitor.php:65]: instead of relying on a callback we accumulate class description here

File: src/CLI/Check.php [hunk 2]
```diff
@@ -41,7 +41,12 @@ protected function configure(): void
     protected function execute(InputInterface $input, OutputInterface $output): int
     {
         ini_set('memory_limit', '-1');
+
         try {
+            $verbose = $input->getOption('verbose');
+
+            $progress = $verbose ? new DebugProgress($output) : new ProgressBarProgress($output);
```
Comment (micheleorselli) [inline src/CLI/Check.php:48]: DebugProgress shows every file parsed
ProgressBarOutput shows a progress bar

File: src/CLI/Progress.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?php
+declare(strict_types=1);
+
+namespace Arkitect\CLI;
+
+use Arkitect\ClassSet;
+
+interface Progress
```
Comment (micheleorselli) [inline src/CLI/Progress.php:8]: every steps method in this interface represents a different steps in analysis. Concrete implementations can decide which information to show and how

File: src/Analyzer/FileVisitor.php (hunk 1 above)
Comment (AlessandroMinoccheri) [reply src/Analyzer/FileVisitor.php:65]: Make sense this approach

**Comment by AlessandroMinoccheri** [comment]
For me, the approach and the code are correct.
I think it could be useful to add a little section or a description of the new option `-v` inside the README file.

**Comment by micheleorselli** [comment]
> For me, the approach and the code are correct.
> I think it could be useful to add a little section or a description of the new option `-v` inside the README file.

yup, I'll proceed to complete the cleanup and add some info on the README

File: src/CLI/ProgressBarProgress.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    public function endParsingFile(string $file): void
+    {
+        $this->progress->advance();
+    }
+
+    public function endFileSetAnalysis(ClassSet $set): void
+    {
+        $this->progress->finish();
+        $this->output->writeln('');
+        $this->output->writeln('');
```
Comment (AlessandroMinoccheri) [inline src/CLI/ProgressBarProgress.php:43]: Are necessary these empty writelines?

Comment (micheleorselli) [reply src/CLI/ProgressBarProgress.php:43]: Without them, any output would be in the same line of the progress bar

Comment (AlessandroMinoccheri) [reply src/CLI/ProgressBarProgress.php:43]: Ok, now it's clear!

**Review by AlessandroMinoccheri — APPROVED** [review]
LGTM 👍

**Pull request merged**